graph/src/blockchain/block_stream.rs

```diff
@@ -6,10 +6,33 @@ use thiserror::Error;
 use super::{Block, BlockPtr, Blockchain};
 use crate::components::store::BlockNumber;
 use crate::firehose::bstream;
+use crate::prelude::tokio::sync::mpsc::Sender;
 use crate::{prelude::*, prometheus::labels};
 
+pub async fn stream_to_channel<C: Blockchain>(
+    stream: Box<dyn BlockStream<C>>,
+    guard: Arc<SharedCancelGuard>,
+    sender: Sender<Result<BlockStreamEvent<C>, CancelableError<Error>>>,
+) -> Result<(), Error> {
```

Comment:
I think it would be more encapsulated and require less changes to the instance manager if this looks like
```rust
pub async fn spawn_buffered(stream: Box<dyn BlockStream<C>>, buffer_size: usize) -> Box<dyn BlockStream<C>>
```
This would internally spawn a task, create a channel, and wrap the receiver in a stream, creating a new block stream. This does not need to handle cancellation, the cancellation code in the instance manager can remain as it is today, which means the receiver will be canceled rather than the block stream directly, but that's fine since ultimately that should cause the block stream task to terminate.

Reply:
generally speaking I think it makes sense to leave concurrency handling to caller because it improves testing and is more flexible. 

This also doesn't work very well because we need ways of cancelling which means we actually need to pass Cancellable everywhere. I had some issues getting those types to line up and play nice with async. 

My preference would be to prioritise the changes to be testable over being "concise to use"

Reply:
On testability, I agree that a function which spawns a task is harder to test. But as we discussed we can have the best of both options by keeping `stream_to_channel`, for testability, and have `spawn_buffered` call it so we can have a cleaner interface and avoiding pushing this extra bit of complexity directly into the instance manager.

On cancellation, as we discussed cancelling the receiver should be fine.